### PR TITLE
Fixes #25977 - Use host info provider for variables

### PR DESCRIPTION
**Pull request by xprazak2**

### CODE REVIEW

**Comment by xprazak2** [comment]
This should also handle nested values for variables correctly.

**Comment by ares** [comment]
Now the output ends in root level, e.g.
```
foreman_ansible:
  a: tada
parameters:
  kt_env: 
  kt_cv: 
```
which is no longer valid ENC output for puppet. Previously it was merged under parameters which I believe end in the ansible inventory.

```
parameters:
  kt_env: 
  kt_cv: 
  a: tada
```

So I suppose we need to wrap it in parameters? The `foreman_ansible` prefix should be added based on settings.

**Comment by xprazak2** [comment]
Updated, I added a new setting that determines whether to add namespace to ENC. It has no impact on variables in inventory file.

**Comment by xprazak2** [comment]
Updated, I removed the setting for namespace in ENC.

**Comment by ares** [comment]
test failures seems related

**Comment by ares** [comment]
Sorry I meant unrelated, it only needs to be rebased on current master I guess.

So with this patch I think the behavior changed (but I think in a good direction), we now add all ansible variables to inventory and ENC, their default value is used if no matcher is matching. That is good but user has no way to keep default for ansible, we're missing "override" checkbox we have in smart class parameters.

I would be fine with merging this as is, if you could also add "override" flag in separate PR. Based on today lunch discussion, it shouldn't be too complicated.

**Comment by xprazak2** [comment]
I pushed a second commit that allows us to use `override` and `omit` with ansible variables.

Then I realized omit will need additional changes.

**Comment by xprazak2** [comment]
`omit` will not work without additional changes in core, but having `override` should be enough for now.

I also disabled editing of the ansible variable name in form, I am not sure if it was enabled intentionally.

File: app/services/foreman_ansible/ansible_info.rb [hunk 1]
```diff
@@ -10,7 +10,7 @@ def ansible_params
 
       variables.each_with_object({}) do |var, memo|
         value = values[var]
-        memo[var.key] = value if value
+        memo[var.key] = value if value && var.override
```
Comment (ares) [inline app/services/foreman_ansible/ansible_info.rb]: could we use override few lines above when we search for AnsibleVariables? or do we need to load all in order to correctly compute values_has?

File: app/views/ansible_variables/_fields.erb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    <%= text_f f, :key, :disabled => true, :size => "col-md-8" %>
+    <%= f.hidden_field :key %>
+    <%= textarea_f f, :description, :rows => :auto, :size => "col-md-8", :class => "no-stretch" %>
+    <%= select_f f, :ansible_role_id, [f.object.ansible_role], :id, :to_label, {}, :label => _("Ansible Role"), :disabled => true %>
+  </fieldset>
+  <fieldset>
+    <h2><%= _("Default Behavior") %></h2>
+    <h6><%= _("Override the default value of the Ansible variable.") %></h6>
+    <%= checkbox_f(f, :override, :onchange => 'toggleOverrideValue(this)', :size => "col-md-8",
+                   :label_help => _('Whether the Ansible variable value is managed by Foreman.')
```
Comment (ares) [inline app/views/ansible_variables/_fields.erb]: Could we describe this a bit more? Something like:

"Mark the variable to be managed by Foreman. When the Ansible role of this variable is assigned to a host, the default value will be added to Ansible inventory as a host variable. To set a different value for such variable, specify matchers."

**Comment by ares** [comment]
oh actually one more thing, we should highlight the overriden somehow in the table, either a new column or the flag we use for puppet

**Comment by xprazak2** [comment]
I search only for overriden variables, changed the label text and added flags:

overriden-vars

**Comment by san7ket** [comment]
@xprazak2 I don't know what I am missing, but the ansible variables are all empty for me, I mean there is no default value set in them. Could you point me in right direction please

**Comment by ares** [comment]
@san7ket we don't import the default value, once you import the variable, you need to go to its edit form and check "override". Once you assign this variable's role to the host, the configured default value is set. They can be overridden using matchers at the bottom of the form. Hope that helps

**Comment by ares** [comment]
not sure if this is caused by the parameter types introduced in core or some change here, but the variable type selector is now visually broken 
vars

**Comment by xprazak2** [comment]
Yes, caused by changes in core, should be fixed now. 

I also opened https://github.com/theforeman/foreman/pull/6506 to enable/disable type field based on override checkbox.

**Comment by ares** [comment]
[test foreman_ansible]

**Comment by ares** [comment]
[test foreman_ansible]

**Comment by ares** [comment]
triggered manually, the comment does not seem to work http://ci.theforeman.org/job/foreman_ansible-pull-request/287/

**Comment by ares** [comment]
[test foreman_ansible]

**Comment by ares** [comment]
APJ

**Comment by xprazak2** [comment]
```
rake aborted!
ActiveRecord::AdapterNotSpecified: 'production' database is not configured. Available: ["development", "test"]

```

**Comment by ares** [comment]
Failures on sqlite does not seem related to this PR, thanks @xprazak2, merging!

**Pull request merged**